### PR TITLE
Set up FTP server

### PR DESCRIPTION
**Pull request by andrewsanchez**

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -51,4 +51,18 @@ jobs:
...
+        ansible-playbook \
+        -i inventory.yml \
+        --private-key=key.pem \
+        -u ubuntu \
+        --become \
+        -e FTP_USER_NAME=$FTP_USER_NAME \
+        -e FTP_USER_PASS=$FTP_USER_PASS \
+        -e FTP_USER_HOME=$FTP_USER_HOME \
+        provision.yml
+
```
Comment (thermokarst) [inline .github/workflows/deploy.yml]: Extra newline

```suggestion
```

File: provision.yml [hunk 2]
```diff
@@ -2,7 +2,10 @@
 
 - name: set up web services
   hosts: all
+  vars:
+    ftp_dir: "ftp"
```
Comment (thermokarst) [inline provision.yml]:
```suggestion
    ftp_dir: ftp
```

File: roles/ftp/templates/ftp_server.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+services:
+  server:
+    build: .
+    image: q2-ftp:latest
+    privileged: true
+    volumes:
+      - ./data:/home/ftpusers
+    ports:
+      - "2121:21"
+      - "30000-30099:30000-30099"
```
Comment (thermokarst) [inline roles/ftp/templates/ftp_server.yml:12]: We will need to update the ufw rules (in the ufw role) to expose ports 21 and 30000-30099

File: roles/ftp/templates/ftp_server.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+version: '3'
+services:
+  server:
+    build: .
+    image: q2-ftp:latest
```
Comment (thermokarst) [inline roles/ftp/templates/ftp_server.yml]:
```suggestion
    image: library-ftp:latest
```

File: roles/ftp/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+---
+
+- name: make ftp dir
+  file:
+    name: "{{ item }}"
+    state: directory
+  with_items:
+    - "{{ ftp_dir }}"
+    - "{{ ftp_dir }}/data"
```
Comment (thermokarst) [inline roles/ftp/tasks/main.yml]: I need to take a look at the default FS layout of the conda-channel-resource, as well as the existing nginx server, to update this accordingly.

File: roles/ftp/templates/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,7 @@
+FROM stilliard/pure-ftpd:hardened
+
+ENV FTP_MAX_CLIENTS=50
+
+CMD /run.sh -c 75 -C 20 -l puredb:/etc/pure-ftpd/pureftpd.pdb -E -j -R -P $PUBLICHOST -p 30000:30099 -d -d -O w3c:/var/log/pure-ftpd/transfer.log --tls=0
```
Comment (thermokarst) [inline roles/ftp/templates/Dockerfile]: We probably should think about TLS. Alternatively, we can set up an IP list to only allow incoming connections from one of our NAU IP addresses.

File: roles/ftp/tasks/main.yml (hunk 5 above)
Comment (thermokarst) [reply roles/ftp/tasks/main.yml]: The easiest solution would be to set up a symbolic link, so once we sort out what needs to be linked where, we can add a step to this set of tasks.

File: roles/ftp/templates/ftp_server.yml (hunk 3 above)
Comment (andrewsanchez) [reply roles/ftp/templates/ftp_server.yml:12]: Does `roles/ufw/tasks/main.yml` look ok now?

File: roles/ftp/templates/ftp_server.yml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+version: '3'
+services:
+  server:
+    build: .
+    image: library-ftp:latest
```
Comment (thermokarst) [inline roles/ftp/templates/ftp_server.yml]:
```suggestion
    image: library-staging-ftp:latest
```

File: roles/ftp/tasks/main.yml (hunk 5 above)
Comment (thermokarst) [inline roles/ftp/tasks/main.yml]:
```suggestion
    - "{{ ftp_dir }}/tls"
```

File: roles/ftp/templates/ftp_server.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+version: '3'
+services:
+  server:
+    build: .
+    image: library-staging-ftp:latest
+    privileged: true
+    volumes:
+      - ./data:/home/ftpusers
```
Comment (thermokarst) [inline roles/ftp/templates/ftp_server.yml]:
```suggestion
      - /usr/share/nginx/html/packages.qiime2.org/qiime2:/home/ftpusers
      - {{ ftp_dir }}/tls:/etc
```

File: roles/ftp/templates/Dockerfile (hunk 6 above)
Comment (thermokarst) [inline roles/ftp/templates/Dockerfile]:
```suggestion
CMD /run.sh \
  -c 75 \
  -C 20 \
  -l puredb:/etc/pure-ftpd/pureftpd.pdb \
  -E \
  -j \
  -R \
  -P $PUBLICHOST \
  -p 30000:30099 \
  -O w3c:/var/log/pure-ftpd/transfer.log \
  --tls=2
```

File: roles/ftp/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
```
Comment (thermokarst) [inline roles/ftp/tasks/main.yml:10]:
```suggestion

- name: copy qiime2.org certs
  copy:
    src: "{{ item.src }}"
    dest: {{ ftp_dir }}/tls/{{ item.dest }}
  with_items:
    - { src: "*.qiime2.org.chained.crt", dest: pure-ftpd.pem }
    - { src: "*.qiime2.org.key", dest: pure-ftpd.key }
```

File: roles/ftp/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    - "{{ ftp_dir }}/tls"
+
+- name: copy qiime2.org certs
+  copy:
+    src: "{{ item.src }}"
+    dest: {{ ftp_dir }}/tls/{{ item.dest }}
+  with_items:
+    - { src: "*.qiime2.org.chained.crt", dest: pure-ftpd.pem }
+    - { src: "*.qiime2.org.key", dest: pure-ftpd.key }
+- name: copy ftp server docker compose template
```
Comment (thermokarst) [inline roles/ftp/tasks/main.yml:18]:
```suggestion
    - { src: "*.qiime2.org.key", dest: pure-ftpd.key }
    
- name: copy ftp server docker compose template
```

File: provision.yml [hunk 11]
```diff
@@ -2,7 +2,11 @@
 
 - name: set up web services
   hosts: all
+  vars:
+    ftp_dir: ftp
```
Comment (thermokarst) [inline provision.yml:7]:
```suggestion
  # TODO Remove when Busywork is decommissioned
  hosts: all
  vars:
    ftp_dir: ftp
```

File: provision.yml [hunk 12]
```diff
@@ -2,7 +2,11 @@
...
   roles:
     - ufw
     - nginx
     - docker
+    # TODO Remove when Busywork is decomissioned
```
Comment (thermokarst) [inline provision.yml]:
```suggestion
    # TODO Remove when Busywork is decommissioned
```

File: .github/workflows/deploy.yml [hunk 13]
```diff
@@ -51,4 +51,17 @@ jobs:
 
     - name: provision server(s)
       if: github.ref == 'refs/heads/main'
-      run: ansible-playbook -i inventory.yml --private-key=key.pem -u ubuntu --become provision.yml
+      env:
+        FTP_USER_NAME: "${{ secrets.FTP_USER_NAME }}"
+        FTP_USER_PASS: "${{ secrets.FTP_USER_PASS }}"
+        FTP_USER_HOME: "${{ secrets.FTP_USER_HOME }}"
```
Comment (thermokarst) [inline .github/workflows/deploy.yml:58]:
```suggestion
      # TODO Remove when Busywork is decommissioned (and remove `-e` below)
      env:
        FTP_USER_NAME: "${{ secrets.FTP_USER_NAME }}"
        FTP_USER_PASS: "${{ secrets.FTP_USER_PASS }}"
        FTP_USER_HOME: "${{ secrets.FTP_USER_HOME }}"
```

File: roles/ufw/tasks/main.yml [hunk 14]
```diff
@@ -7,6 +7,8 @@
     port: "{{ item }}"
     proto: tcp
   with_items:
+    - "21"
     - "22"
     - "80"
     - "443"
+    - "30000:30099"
```
Comment (thermokarst) [inline roles/ufw/tasks/main.yml]:
```suggestion
  with_items:
    - "21"  # TODO Remove when Busywork is decommissioned
    - "22"
    - "80"
    - "443"
    - "30000:30099"  # TODO Remove when Busywork is decommissioned
```

File: roles/ftp/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    name: "{{ item }}"
+    state: directory
+  with_items:
+    - "{{ ftp_dir }}"
+    - "{{ ftp_dir }}/tls"
+
+- name: copy qiime2.org certs
+  copy:
+    src: "{{ item.src }}"
+    dest: {{ ftp_dir }}/tls/{{ item.dest }}
```
Comment (thermokarst) [inline roles/ftp/tasks/main.yml]:
```suggestion
    dest: "{{ ftp_dir }}/tls/{{ item.dest }}"
```

**Review by thermokarst — APPROVED** [review]

File: roles/ftp/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    - "{{ ftp_dir }}/tls"
+
+- name: copy qiime2.org certs
+  copy:
+    src: "{{ item.src }}"
+    dest: "{{ ftp_dir }}/tls/{{ item.dest }}"
+  with_items:
+    - { src: "*.qiime2.org.chained.crt", dest: pure-ftpd.pem }
+    - { src: "*.qiime2.org.key", dest: pure-ftpd.key }
+
```
Comment (thermokarst) [inline roles/ftp/tasks/main.yml]:
```suggestion

```

**Pull request merged**